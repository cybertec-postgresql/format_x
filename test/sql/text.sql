CREATE EXTENSION IF NOT EXISTS format_x;

select format_x(NULL);
select format_x('Hello');
select format_x('Hello %s', 'World');
select format_x('Hello %%');
select format_x('Hello %%%%');
-- should fail
select format_x('Hello %s %s', 'World');
select format_x('Hello %s');
select format_x('Hello %x', 20);
-- check literal and sql identifiers
select format_x('INSERT INTO %I VALUES(%L,%L)', 'mytab', 10, 'Hello');
select format_x('%s%s%s','Hello', NULL,'World');
select format_x('INSERT INTO %I VALUES(%L,%L)', 'mytab', 10, NULL);
select format_x('INSERT INTO %I VALUES(%L,%L)', 'mytab', NULL, 'Hello');
-- should fail, sql identifier cannot be NULL
select format_x('INSERT INTO %I VALUES(%L,%L)', NULL, 10, 'Hello');
-- check positional placeholders
select format_x('%1$s %3$s', 1, 2, 3);
select format_x('%1$s %12$s', 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12);
-- should fail
select format_x('%1$s %4$s', 1, 2, 3);
select format_x('%1$s %13$s', 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12);
select format_x('%0$s', 'Hello');
/* select format_x('%*0$s', 'Hello'); */
select format_x('%1$', 1);
select format_x('%1$1', 1);
-- check mix of positional and ordered placeholders
select format_x('Hello %s %1$s %s', 'World', 'Hello again');
select format_x('Hello %s %s, %2$s %2$s', 'World', 'Hello again');
-- check variadic labeled arguments
select format_x('%s, %s', variadic array['Hello','World']);
select format_x('%s, %s', variadic array[1, 2]);
select format_x('%s, %s', variadic array[true, false]);
select format_x('%s, %s', variadic array[true, false]::text[]);
-- check variadic with positional placeholders
select format_x('%2$s, %1$s', variadic array['first', 'second']);
select format_x('%2$s, %1$s', variadic array[1, 2]);
-- variadic argument can be array type NULL, but should not be referenced
select format_x('Hello', variadic NULL::int[]);
-- variadic argument allows simulating more than FUNC_MAX_ARGS parameters
select format_x(string_agg('%s',','), variadic array_agg(i))
from generate_series(1,200) g(i);
-- check field widths and left, right alignment
select format_x('>>%10s<<', 'Hello');
select format_x('>>%10s<<', NULL);
select format_x('>>%10s<<', '');
select format_x('>>%-10s<<', '');
select format_x('>>%-10s<<', 'Hello');
select format_x('>>%-10s<<', NULL);
select format_x('>>%1$10s<<', 'Hello');
select format_x('>>%1$-10I<<', 'Hello');
/* select format_x('>>%2$*1$L<<', 10, 'Hello'); */
/* select format_x('>>%2$*1$L<<', 10, NULL); */
/* select format_x('>>%2$*1$L<<', -10, NULL); */
/* select format_x('>>%*s<<', 10, 'Hello'); */
/* select format_x('>>%*1$s<<', 10, 'Hello'); */
select format_x('>>%-s<<', 'Hello');
select format_x('>>%10L<<', NULL);
/* select format_x('>>%2$*1$L<<', NULL, 'Hello'); */
/* select format_x('>>%2$*1$L<<', 0, 'Hello'); */
