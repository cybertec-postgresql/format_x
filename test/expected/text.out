CREATE EXTENSION IF NOT EXISTS format_x;
NOTICE:  extension "format_x" already exists, skipping
select format_x(NULL);
 format_x 
----------
 
(1 row)

select format_x('Hello');
 format_x 
----------
 Hello
(1 row)

select format_x('Hello %s', 'World');
  format_x   
-------------
 Hello World
(1 row)

select format_x('Hello %%');
 format_x 
----------
 Hello %
(1 row)

select format_x('Hello %%%%');
 format_x 
----------
 Hello %%
(1 row)

-- should fail
select format_x('Hello %s %s', 'World');
ERROR:  too few arguments for format_x()
select format_x('Hello %s');
ERROR:  too few arguments for format_x()
select format_x('Hello %x', 20);
ERROR:  unrecognized format_x() type specifier "x"
HINT:  For a single "%" use "%%".
-- check literal and sql identifiers
select format_x('INSERT INTO %I VALUES(%L,%L)', 'mytab', 10, 'Hello');
                format_x                
----------------------------------------
 INSERT INTO mytab VALUES('10','Hello')
(1 row)

select format_x('%s%s%s','Hello', NULL,'World');
  format_x  
------------
 HelloWorld
(1 row)

select format_x('INSERT INTO %I VALUES(%L,%L)', 'mytab', 10, NULL);
              format_x               
-------------------------------------
 INSERT INTO mytab VALUES('10',NULL)
(1 row)

select format_x('INSERT INTO %I VALUES(%L,%L)', 'mytab', NULL, 'Hello');
                format_x                
----------------------------------------
 INSERT INTO mytab VALUES(NULL,'Hello')
(1 row)

-- should fail, sql identifier cannot be NULL
select format_x('INSERT INTO %I VALUES(%L,%L)', NULL, 10, 'Hello');
ERROR:  null values cannot be formatted as an SQL identifier
-- check positional placeholders
select format_x('%1$s %3$s', 1, 2, 3);
 format_x 
----------
 1 3
(1 row)

select format_x('%1$s %12$s', 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12);
 format_x 
----------
 1 12
(1 row)

-- should fail
select format_x('%1$s %4$s', 1, 2, 3);
ERROR:  too few arguments for format_x()
select format_x('%1$s %13$s', 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12);
ERROR:  too few arguments for format_x()
select format_x('%0$s', 'Hello');
ERROR:  format specifies argument 0, but arguments are numbered from 1
select format_x('%*0$s', 'Hello');
ERROR:  format specifies argument 0, but arguments are numbered from 1
select format_x('%1$', 1);
ERROR:  unterminated format_x() type specifier
HINT:  For a single "%" use "%%".
select format_x('%1$1', 1);
ERROR:  unterminated format_x() type specifier
HINT:  For a single "%" use "%%".
-- check mix of positional and ordered placeholders
select format_x('Hello %s %1$s %s', 'World', 'Hello again');
           format_x            
-------------------------------
 Hello World World Hello again
(1 row)

select format_x('Hello %s %s, %2$s %2$s', 'World', 'Hello again');
                     format_x                     
--------------------------------------------------
 Hello World Hello again, Hello again Hello again
(1 row)

-- check variadic labeled arguments
select format_x('%s, %s', variadic array['Hello','World']);
   format_x   
--------------
 Hello, World
(1 row)

select format_x('%s, %s', variadic array[1, 2]);
 format_x 
----------
 1, 2
(1 row)

select format_x('%s, %s', variadic array[true, false]);
 format_x 
----------
 t, f
(1 row)

select format_x('%s, %s', variadic array[true, false]::text[]);
  format_x   
-------------
 true, false
(1 row)

-- check variadic with positional placeholders
select format_x('%2$s, %1$s', variadic array['first', 'second']);
   format_x    
---------------
 second, first
(1 row)

select format_x('%2$s, %1$s', variadic array[1, 2]);
 format_x 
----------
 2, 1
(1 row)

-- variadic argument can be array type NULL, but should not be referenced
select format_x('Hello', variadic NULL::int[]);
 format_x 
----------
 Hello
(1 row)

-- variadic argument allows simulating more than FUNC_MAX_ARGS parameters
select format_x(string_agg('%s',','), variadic array_agg(i))
from generate_series(1,200) g(i);
                                                                                                                                                                                                                                                                                                                                                      format_x                                                                                                                                                                                                                                                                                                                                                       
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200
(1 row)

-- check field widths and left, right alignment
select format_x('>>%10s<<', 'Hello');
    format_x    
----------------
 >>     Hello<<
(1 row)

select format_x('>>%10s<<', NULL);
    format_x    
----------------
 >>          <<
(1 row)

select format_x('>>%10s<<', '');
    format_x    
----------------
 >>          <<
(1 row)

select format_x('>>%-10s<<', '');
    format_x    
----------------
 >>          <<
(1 row)

select format_x('>>%-10s<<', 'Hello');
    format_x    
----------------
 >>Hello     <<
(1 row)

select format_x('>>%-10s<<', NULL);
    format_x    
----------------
 >>          <<
(1 row)

select format_x('>>%1$10s<<', 'Hello');
    format_x    
----------------
 >>     Hello<<
(1 row)

select format_x('>>%1$-10I<<', 'Hello');
    format_x    
----------------
 >>"Hello"   <<
(1 row)

select format_x('>>%2$*1$L<<', 10, 'Hello');
    format_x    
----------------
 >>   'Hello'<<
(1 row)

select format_x('>>%2$*1$L<<', 10, NULL);
    format_x    
----------------
 >>      NULL<<
(1 row)

select format_x('>>%2$*1$L<<', -10, NULL);
    format_x    
----------------
 >>NULL      <<
(1 row)

select format_x('>>%*s<<', 10, 'Hello');
    format_x    
----------------
 >>     Hello<<
(1 row)

select format_x('>>%*1$s<<', 10, 'Hello');
    format_x    
----------------
 >>     Hello<<
(1 row)

select format_x('>>%-s<<', 'Hello');
 format_x  
-----------
 >>Hello<<
(1 row)

select format_x('>>%10L<<', NULL);
    format_x    
----------------
 >>      NULL<<
(1 row)

select format_x('>>%2$*1$L<<', NULL, 'Hello');
  format_x   
-------------
 >>'Hello'<<
(1 row)

select format_x('>>%2$*1$L<<', 0, 'Hello');
  format_x   
-------------
 >>'Hello'<<
(1 row)

